nixos/modules/services/web-apps/keycloak.nix

```diff
@@ -0,0 +1,441 @@
+                    echo "Reassigning ownership of database keycloak to user ${cfg.databaseUsername} failed on last boot. Failing..." >&2
+                    exit 1
+                fi
+                touch "${config.services.postgresql.dataDir}/.reassigning_keycloak"
+                $PSQL "keycloak" -tAc "REASSIGN OWNED BY \"$current_owner\" TO \"${cfg.databaseUsername}\""
+                rm "${config.services.postgresql.dataDir}/.reassigning_keycloak"
+            fi
+          '';
+        };
+
+        systemd.services.keycloak = {
+          after = lib.optionals databaseActuallyCreateLocally [
+            "keycloakDatabaseInit.service" "postgresql.service"
+          ];
+          bindsTo = lib.optionals databaseActuallyCreateLocally [
+            "keycloakDatabaseInit.service" "postgresql.service"
+          ];
+          wantedBy = [ "multi-user.target" ];
+          environment = {
+            JBOSS_LOG_DIR = "/var/log/keycloak";
+            JBOSS_BASE_DIR = "/var/lib/keycloak";
+            JBOSS_MODULEPATH = "${cfg.package}/modules";
+          };
+          serviceConfig = {
+            ExecStartPre = systemdPrefixScript "startPre" "+" (''
+              set -eu
+
+              if ! [[ -d /var/lib/keycloak/configuration ]]; then
+                  cp -r ${cfg.package}/standalone/configuration /var/lib/keycloak/
+              fi
+
+              if ! [[ -e ${cfg.databasePasswordFile} ]]; then
+                  echo "Database password file ${cfg.databasePasswordFile} not found! Aborting." >&2
+                  exit 1
+              fi
+              cp ${keycloakConfig} /var/lib/keycloak/configuration/standalone.xml
+              db_password="$(<'${cfg.databasePasswordFile}')"
+              ${pkgs.replace}/bin/replace-literal -fe '@db-password@' "$db_password" /var/lib/keycloak/configuration/standalone.xml
+
+              mkdir -p /var/lib/keycloak/{deployments,ssl}
+
+              '' + lib.optionalString (cfg.certificatePrivateKeyBundle != null) ''
+              if [[ -f "${cfg.certificatePrivateKeyBundle}" ]]; then
+                  pushd /var/lib/keycloak/ssl/
+                  cat "${cfg.certificatePrivateKeyBundle}" /etc/ssl/certs/ca-certificates.crt > allcerts.pem
+                  ${pkgs.openssl}/bin/openssl pkcs12 -export -in ${cfg.certificatePrivateKeyBundle} -chain \
+                                                     -name "${cfg.frontendUrl}" -out certificate_private_key_bundle.p12 \
+                                                     -CAfile allcerts.pem -passout pass:notsosecretpassword
+                  popd
+              fi
+              '' + ''
+              chown -R keycloak:keycloak /var/lib/keycloak/
+              chmod -R u+rwX,g-rwx,o-rwx /var/lib/keycloak/
+              ${pkgs.acl}/bin/setfacl -R -m d:u:keycloak:rwX /var/lib/keycloak/
+              ${pkgs.acl}/bin/setfacl -R -m d:g::-,d:o::- /var/lib/keycloak/
+              chown keycloak:keycloak /run/wrappers/bin/keycloak
```

Comment:
Possibly it is obvious... but I'm not following the purpose of this.

Reply:
Well, it was a quick and dirty way to assign the correct permissions for current and future files and the wrapper was needed when I, for some reason, decided to use `CapabilityBoundingSet` instead of `AmbientCapabilities`. All of it is removed now :)